docs/v1/P4Runtime-Spec.mdk

```diff
@@ -3293,10 +3306,68 @@ direct resources that are being queried.  Idle timeouts are not supported for
 static entries. If the table requires a priority value for entries, the server
 must populate the `priority` field appropriately, starting at 1 for the lowest
 priority entry and incrementing the value by 1 for each successive entry. Note
-that P4~16~ does not support assigning explicit priorities to static
-entries. When a priority value is required (&eg; for tables including `RANGE`,
-`TERNARY` or `OPTIONAL` matches), it is inferred based on the order in which
-entries appear in the table declaration.
+that P4~16~ does not support assigning explicit priorities to entries
+declared with `const entries`[^ConstEntriesPriorities].  When a
+priority value is required (&eg; for tables including `RANGE`,
+`TERNARY` or `OPTIONAL` matches), it is inferred based on the order in
+which entries appear in the table declaration.
+
+[^ConstEntriesPriorities]: This is not yet explicit in the P4~16~
+    language specification, but will become so if this or a similar
+    clarification is added to it:
+    <https://github.com/p4lang/p4-spec/pull/1259>
+
+### Tables with initial entries { #sec-tables-with-initial-entries }
+
+Tables with initial entries are those identified by the table property
+`entries` in the P4~16~ source code (with no `const` qualifier before
+`entries`).  In the P4Info, such tables have `has_initial_entries`
+flag equal to true, but `is_const_table` is false.  For tables
+declared with `const entries`, see Section [#sec-constant-tables].
+
+For every P4 table, it will fall into one of these three categories:
+
+* Neither `entries` nor `const entries` are declared in the source
+  code, and thus `is_const_table` and `has_initial_entries` will both
+  be false.  This will be called a "normal table" in this section
+  below.
+* The table has `const entries` declared, and thus a separate
+  `entries` property is not permitted by the language.  Such a table
+  will have `is_const_table` true and `has_initial_entries` true.
```

Comment:
P4 supports declaring a table with `const entries`, or with `entries` (without `const` before it), with an empty list of entries.

Reply:
So technically, we might have a 4th category of tables with `is_const_table = true`  and `has_initial_entries = false`?

Reply:
Elsewhere in these changes the definition of `has_initial_entries` true is that the P4 source code has the `entries` property present for the table, vs. not present for the table.  Do you think that is a reasonable definition?

Reply:
Maybe to be more explicit, are you saying that you think `has_initial_entries` should be true if and only if there is an `entries` property for the table in the P4 source code, and there is at least one entry listed there?  So if the list contains 0 entries, `has_initial_entries` should be false?

Reply:
@smolkaj I thought about this some more since my last comment, and came to the conclusion that it does seem to make more sense if `has_initial_entries` is true if and only if the P4 source code has an `entries` table property, _and_ the list of entries is not empty.  With commit 19 on this PR, I believe I have made all changes required to consistently define it this way.

This does mean that a normal table is either:
(a) one that has neither `entries` nor `const entries` properties in the source code
(b) one that has `entries = { }`, with no `const` before `entries`

Two tables declared as (a) or (b) would both have `is_const_table` false and `has_initial_entries` false as proposed by commit 19 version of this PR.  That makes a kind of sense to me, since the things that the client can do to either a table declared like (a) or (b) would be exactly the same, i.e. there are no entries initially loaded into the table when you load the P4 program, and the runtime can insert, modify, or delete any entries it wants to at run time.

Reply:
> Elsewhere in these changes the definition of has_initial_entries true is that the P4 source code has the entries property present for the table, vs. not present for the table. Do you think that is a reasonable definition?

I hadn't thought about that, and actually do think this is also a reasonable definition (even if the `entries` property specifies the empty set of entries, which is still a set of entries).

> are you saying that you think has_initial_entries should be true if and only if there is an entries property for the table in the P4 source code, and there is at least one entry listed there? So if the list contains 0 entries, has_initial_entries should be false?

I was just asking a question -- sorry if it sounded like I had thought this through to the end. 

> I thought about this some more since my last comment, and came to the conclusion that it does seem to make more sense if has_initial_entries is true if and only if the P4 source code has an entries table property, and the list of entries is not empty.

This also SGTM. Just to check my understanding, the main difference is that we now have "constant tables" with and without `has_initial_entries`.

Actually, it does seem nice to allow a constant table with `has_initial_entries: false`, since clearly this is easier to support by the switch than if `has_initial_entries: true`.

> (b) one that has entries = { }, with no const before entries

Orthogonal to this PR, but perhaps we should ask p4c to disallow this? One weird corner case less?

Reply:
> This also SGTM. Just to check my understanding, the main difference is that we now have "constant tables" with and without `has_initial_entries`.

Yes, and the text I have in this PR now mentions both of these possibilities explicitly, and calls both variants "constant tables".

> Orthogonal to this PR, but perhaps we should ask p4c to disallow this? One weird corner case less?

I suppose we could, but it doesn't seem that bad to me if `entries = { }` has the same meaning as a table with no `entries` table property declared in the source code at all.